File: android/src/main/java/com/reactnativecommunity/viewpager/FragmentAdapter.java
package com.reactnativecommunity.viewpager;

import android.view.View;
import android.view.ViewParent;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.viewpager2.adapter.FragmentStateAdapter;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class FragmentAdapter extends FragmentStateAdapter {
    private final BiMap<String, Integer> mChildKeyToPosition = new BiMap<>();
    private final List<WeakReference<ViewPagerFragment>> mFragments = new LinkedList<>();
    private final List<View> mReactChildrenViews = new ArrayList<>();
    private final List<String> mUnprocessedChildKeys = new ArrayList<>();
    private int count = 0;
    private int offset = 0;

    private int appliedCount = 0;
    private String currentKey = null;
    private int currentPosition = 0;

    private int currentTransactionId = 0;

    public FragmentAdapter(@NonNull FragmentActivity fragmentActivity) {
        super(fragmentActivity);
    }

    @NonNull
    @Override
    public Fragment createFragment(int position) {
        ViewPagerFragment fragment = new ViewPagerFragment(position, getViewAtPosition(position));
        mFragments.add(new WeakReference<>(fragment));
        return fragment;
    }

    @Override
    public int getItemCount() {
        return count;
    }

    @Override
    public long getItemId(int position) {
        View reactView = getViewAtPosition(position);
        return reactView == null ? super.getItemId(position) : reactView.getId();
    }

    @Override
    public boolean containsItem(long itemId) {
        for (View reactView : mReactChildrenViews) {
            if (reactView.getId() == itemId) {
                return true;
            }
        }
        return super.containsItem(itemId);
    }

    public void setCount(int count) {
        this.count = count;
    }

    public void setOffset(int offset) {
        this.offset = offset;
    }

    @Nullable
    public View getViewAtPosition(int position) {
        int index = position - offset;
        return index >= 0 && index < mReactChildrenViews.size()
                ? mReactChildrenViews.get(index)
                : null;
    }

    public void onPageSelected(int position) {
        currentKey = mChildKeyToPosition.getByValue(position);
        currentPosition = position;
    }

    public void queueKeyToProcess(String childKey) {
        mUnprocessedChildKeys.add(childKey);
    }

    private void processChildKeys() {
        if (mUnprocessedChildKeys.isEmpty()) {
            return;
        }
        mChildKeyToPosition.clear();
        for (int i = 0; i < mUnprocessedChildKeys.size(); ++i) {
            mChildKeyToPosition.put(mUnprocessedChildKeys.get(i), offset + i);
        }
        mUnprocessedChildKeys.clear();
    }

    public int getTransactionId() {
        return currentTransactionId;
    }

    /**
     * Returns newly selected page index (if there was a change) or null.
     */
    public Integer onAfterUpdateTransaction(int transactionId) {
        if (transactionId != currentTransactionId) {
            // Several sources may trigger this update; coalesce events.
            return null;
        }
        currentTransactionId += 1;

        processChildKeys();

        // Apply structural changes at the beginning of the pager.
        // Note that this may not be the optimal change set, but inaccurate
        // insert/delete locations will be later handled by notifyItemChanged().
        Integer newPosition = null;
        if (currentKey != null) {
            newPosition = mChildKeyToPosition.getByKey(currentKey);
        }
        int delta = 0;
        if (newPosition != null) {
            delta = newPosition - currentPosition;
            if (delta < 0) {
                notifyItemRangeRemoved(0, -delta);
            } else if (delta > 0) {
                notifyItemRangeInserted(0, delta);
            }
        }

        // Apply structural changes at the end of the pager.
        appliedCount += delta;
        if (appliedCount > count) {
            notifyItemRangeRemoved(count, appliedCount - count);
        } else if (appliedCount < count) {
            notifyItemRangeInserted(appliedCount, count - appliedCount);
        }
        appliedCount = count;

        Iterator<WeakReference<ViewPagerFragment>> it = mFragments.iterator();
        while (it.hasNext()) {
            ViewPagerFragment fragment = it.next().get();
            if (fragment == null) {
                it.remove();
            } else if (fragment.onReactViewUpdate(this, delta)) {
                int changedPos = fragment.getPosition();
                if (changedPos >= 0 && changedPos < count) {
                    notifyItemChanged(changedPos);
                }
            }
        }

        return newPosition == null || newPosition == currentPosition ? null : newPosition;
    }

    public void addReactView(View child, int index) {
        mReactChildrenViews.add(index, child);
    }

    public View getReactChildAt(int index) {
        return mReactChildrenViews.get(index);
    }

    public int getReactChildCount() {
        return mReactChildrenViews.size();
    }

    public void removeReactViewAt(int index) {
        View reactView = mReactChildrenViews.remove(index);
        ViewParent rootView = reactView.getParent();
        if (rootView instanceof FrameLayout) {
            ((FrameLayout) rootView).removeAllViews();
        }
    }
}

File: android/src/main/java/com/reactnativepagerview/FragmentAdapter.kt
package com.reactnativepagerview

import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.DiffUtil
import androidx.viewpager2.adapter.FragmentStateAdapter


class FragmentAdapter(fragmentActivity: FragmentActivity) : FragmentStateAdapter(fragmentActivity) {
  private val childrenViews = mutableListOf<View>()
  private var count = 0
  private var offset = 0
  private var isDirty = false
  private val prevItemIds = mutableListOf<Long>()

  override fun createFragment(position: Int): Fragment {
    return ViewPagerFragment(getViewAtPosition(position))
  }

  override fun getItemCount() = count

  override fun getItemId(position: Int): Long {
    val view = getViewAtPosition(position)
    return view?.id?.toLong() ?: UNRENDERED_ID_OFFSET + position
  }

  override fun containsItem(itemId: Long): Boolean {
    if (itemId >= UNRENDERED_ID_OFFSET) {
      val position = itemId - UNRENDERED_ID_OFFSET
      return getViewAtPosition(position.toInt()) == null
    }
    return childrenViews.any { it.id.toLong() == itemId }
  }

  /**
   * Returns true if any changes were applied.
   */
  fun notifyAboutChanges(): Boolean {
    if (!isDirty) {
      return false
    }

    isDirty = false
    val diff = DiffUtil.calculateDiff(
      PagerDiffCallback(prevItemIds, this),
      false
    )
    diff.dispatchUpdatesTo(this)
    return true
  }

  fun setCount(count: Int) {
    if (this.count != count) {
      markDirty()
      this.count = count
    }
  }

  fun setOffset(offset: Int) {
    if (this.offset != offset) {
      markDirty()
      this.offset = offset
    }
  }

  fun addReactView(child: View, index: Int) {
    markDirty()
    childrenViews.add(index, child)
  }

  fun getReactChildAt(index: Int) = childrenViews[index]

  fun getReactChildCount() = childrenViews.size

  fun removeReactViewAt(index: Int) {
    markDirty()
    childrenViews.removeAt(index)
  }

  private fun getViewAtPosition(position: Int): View? {
    val index = position - offset
    return if (index >= 0 && index < childrenViews.size) childrenViews[index] else null
  }

  private fun markDirty() {
    if (isDirty) {
      return
    }
    isDirty = true
    prevItemIds.clear()
    for (position in 0 until itemCount) {
      prevItemIds.add(getItemId(position))
    }
  }

  companion object {
    /**
     * If an id is `UNRENDERED_ID_OFFSET` or higher, it represents a view
     * that is not currently rendered.
     */
    const val UNRENDERED_ID_OFFSET = 0xffffffffL
  }
}
